[#73607174] locksmithctl: Support for multiple endpoints

mattbostock opened the pull request:
Support multiple etcd endpoints to be specified either using commandline
flags or an environment variable.

Maintains the `-endpoint` flag for backwards compatibility. Using
commandline flags, specify `-endpoint=<url>` for each endpoint or
pass a comma-delimited list of endpoints, e.g. `-endpoint=<url>,<url>`.
The former syntax is preferable IMO as it makes more sense semantically
given that `-endpoint` is singular rather than plural.

Specifying multiple endpoints using an environment variable is supported
by passing a comma-delimited list, e.g.
`LOCKSMITHCTL_ENDPOINT=<url>,<url>`.

The default endpoint of `http://127.0.0.1:4001` is retained and prints
correctly when `globalFlagset.Usage()` is called. I used `VisitAll()` to
set the default value so that I could make this distinct to the actual
flag value, which `flag.FlagSet.Var()` doesn't support.

The pattern of using a custom flag type that satisfies the `flag.Value`
interface is borrowed from example 3 of the Go Flag tests:
https://golang.org/src/flag/example_test.go

Fixes #53

* * *

Also, rename `globalFlagset` to `globalFlagSet` for consistency with the
`flag.FlagSet` struct in the Go Flag package:
http://golang.org/pkg/flag/#FlagSet

mattbostock commented:
_bump_ Is there anything I can do to help get this reviewed/merged?

marineam commented:
@jonboulle could you review this please since @crawford is out?

jonboulle commented on locksmithctl/locksmithctl.go:
```diff
@@ -82,20 +96,25 @@ type Command struct {
-	globalFlagset.Parse(os.Args[1:])
-	var args = globalFlagset.Args()
+	globalFlagSet.Parse(os.Args[1:])
+	globalFlagSet.VisitAll(func(f *flag.Flag) {
```
This seems awkward; how about just setting this to the desired value in the `globalFlags` initialisation instead?

jonboulle commented:
The approach seems like a good compromise to me. Thanks for the patch! Just one comment.

mattbostock commented:
I've rebased to amend;

I refactored out the `VisitAll()` function which was setting the default endpoint; instead I've set defined default endpoint as constant and set it [here](https://github.com/mattbostock/locksmith/commit/68eb542cdd33f2d6bf71cbb87fc367dac26607c5?diff=unified#diff-d713ec8c4f7b249df26db9c602fa9bfaR56) and [here](https://github.com/mattbostock/locksmith/commit/68eb542cdd33f2d6bf71cbb87fc367dac26607c5?diff=unified#diff-d713ec8c4f7b249df26db9c602fa9bfaR119) such that it appears in the CLI usage message and is explicitly passed to the etcd client library.

IMHO it's preferable to explicitly determine the default endpoint so that the usage message always matches what the etcd client library will actually use; otherwise the client library could change and the usage message would be incorrect. I'm open to suggestions though, as we could remove lines 119-122 if we don't explicitly set the default and rely on the library default.

jonboulle commented:
@mattbostock That reasoning makes sense.

I would still rather set the default value during init() so we can remove the magic from String(), what do you think about that?

mattbostock commented:
I'm not keen on the magic in `String()` either, though don't think there is a way to set the default during `init()` because `globalFlags.Endpoints` is a custom flag type.

The `Var()` [method](http://golang.org/pkg/flag/#FlagSet.Var) does not allow you to specify a default separate to the actual value; the default is set to whatever `value` argument is passed to `Var()`. Because `globalFlags.Endpoints` is a slice, any value passed in via `Var()` would be appended to the slice; we then have no way of knowing if the first member of the array is the default or has been passed in using a flag.

jonboulle commented:
@mattbostock sorry for the delays, locksmith notifications keep talking through the cracks :-(

What I meant was assigning a specific value to global flags.Endpoints, won't that work? Since the var should still refer to the pointer.

mattbostock commented:
No problem.

If the default value was assigned to `globalFlags.Endpoints` slice, any endpoints passed in explicitly by the user would be appended to that default. So you'd always have `http://127.0.0.1:4001` as an endpoint. Does that answer your question?

jonboulle commented:
Duh, yes, I was totally not thinking about that. Thanks.

mattbostock commented:
Could you please push a new release including this change? Thanks!

crawford commented:
Just tagged 0.2.3. Sorry for the delay.

mattbostock commented:
@crawford Great, thanks!
